src/main/java/meteordevelopment/meteorclient/systems/modules/world/AutoSign.java

```diff
@@ -6,42 +6,293 @@
+        .description("Format string for {time} object.")
+        .defaultValue("HH:mm z")
+        .build()
+    );
+
+    private final Setting<Boolean> utcTimezone = other.add(new BoolSetting.Builder()
+        .name("UTC-timezone")
+        .description("Whether to use UTC or your local timezone.")
+        .defaultValue(true)
+        .build()
+    );
 
     @Override
-    public void onDeactivate() {
-        text = null;
-    }
+    public WWidget getWidget(GuiTheme theme) {
+        WHorizontalList list = theme.horizontalList();
 
-    @EventHandler
-    private void onSendPacket(PacketEvent.Send event) {
-        if (!(event.packet instanceof UpdateSignC2SPacket)) return;
+        WButton starscriptDoc = theme.button("Open Starscript documentation.");
+        starscriptDoc.action = () -> Util.getOperatingSystem().open("https://github.com/MeteorDevelopment/meteor-client/wiki/Starscript");
+        list.add(starscriptDoc);
+
+        WButton simpleDateFormatDoc = theme.button("Open SimpleDateFormat documentation.");
+        simpleDateFormatDoc.action = () -> Util.getOperatingSystem().open("https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html");
+        list.add(simpleDateFormatDoc);
 
-        text = ((UpdateSignC2SPacket) event.packet).getText();
+        return list;
+    }
+
+    public AutoSign() {
+        super(Categories.World, "auto-sign", "Automatically writes signs. Configure what to write dynamically using Starscript placeholders.");
     }
 
     @EventHandler
-    private void onOpenScreen(OpenScreenEvent event) {
-        if (!(event.screen instanceof AbstractSignEditScreen) || text == null) return;
+    private void onOpenScreen(OpenScreenEvent event) throws InterruptedException {
+        if (!(event.screen instanceof AbstractSignEditScreen)) {return;}
+        if (applyingBack) {
+            applyingBack = false;
+            event.cancel();
+            return;
+        }
 
+        // apply custom format and timezone, not the nicest (temporary overwriting global date and time format)
+        try {
+            StandardLib.dateFormat.applyPattern(date.get());
+            StandardLib.timeFormat.applyPattern(time.get());
+        } catch (IllegalArgumentException e) {
+            error(e.getMessage());
+        }
+        if (utcTimezone.get()) {
+            StandardLib.dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+            StandardLib.timeFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+        }
+
+        boolean updated = false;
         SignBlockEntity sign = ((AbstractSignEditScreenAccessor) event.screen).getSign();
 
-        mc.player.networkHandler.sendPacket(new UpdateSignC2SPacket(sign.getPos(), true, text[0], text[1], text[2], text[3]));
+        if (frontEnabled.get()) {
+            boolean written = false;
+            for (Text text : sign.getFrontText().getMessages(false))
+                if (!Objects.equals(text.getString(), "") && !written) {
```

Comment:
`!Objects.equals(text.getString(), "")` can be rewritten as `!text.getString().isEmpty()`